### PR TITLE
Add concurrent control to DistClusterControllerStateModel._controller access to avoid NPE.

### LINKED ISSUE #1752: NPE in the DistClusterControllerStateModel state transition method
### Describe the bug
One NPE is observed in the DistClusterControllerStateModel when cleanup is triggered in parallel with other state transitions.

29066 [HelixTaskExecutor-message_handle_STATE_TRANSITION] ERROR org.apache.helix.messaging.handling.HelixStateTransitionHandler  - Exception while executing a state transition task TestClusterStatusMonitorLifecycle0_2
java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor11.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.helix.messaging.handling.HelixStateTransitionHandler.invoke(HelixStateTransitionHandler.java:386)
	at org.apache.helix.messaging.handling.HelixStateTransitionHandler.handleMessage(HelixStateTransitionHandler.java:313)
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:97)
	at org.apache.helix.messaging.handling.HelixTask.call(HelixTask.java:49)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.apache.helix.participant.DistClusterControllerStateModel.onBecomeLeaderFromStandby(DistClusterControllerStateModel.java:72)
	... 11 more

I believe this exception makes the test TestClusterStatusMonitorLifecycle unstable.

### To Reproduce
Trigger leader to standby transition and close the controller-participant at the same time.

### Expected behavior
No NPE.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Resolves #1752

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

This PR fixes a potential NPE Exception that may be thrown in the DistClusterControllerStateModel state transition methods.
When this error happens, a follower to leader state transition might be interrupted. The controller instance partition will be set with the ERROR state but the controller instance may have connected to the Zookeeper already. This causes inconsistency and leakage since the controller instance won't be properly cleaned up when it is dropped directly from ERROR state.

### Tests

- [ ] The following tests are written for this issue:

The original issue is hard to test. Especially with the fix, it is not possible to reproduce the NPE problem anymore.
I suggest we mainly rely on the TestClusterStatusMonitorLifecycle to validate the correctness. It should become very stable after this change.

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java [hunk 1]
```diff
@@ -288,16 +288,13 @@ public void testClusterStatusMonitorLifecycle() throws Exception {
...
+            exp1);
 
-    // Note, the _asyncTasksThreadPool shutting down logic in GenericHelixController is best effort
-    // there is not guarantee that all threads in the pool is gone. Mossstly they will, but not always.
-    // see https://github.com/apache/helix/issues/1280
     boolean result = TestHelper.verify(() -> ManagementFactory.getPlatformMBeanServer()
         .queryMBeans(new ObjectName("ClusterStatus:*"), exp2).isEmpty(), TestHelper.WAIT_DURATION);
-    Assert.assertTrue(result,
-        "A small chance this may fail due to _asyncThread pool in controller may not shutdown in time. Please check issue 1280 to verify if this is the case.");
+    Assert.assertTrue(result, "Remaining MBeans: " + ManagementFactory.getPlatformMBeanServer()
```
Comment (zhangmeng916) [inline helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java:297]: There are a few other test failures besides this test. Will they be fixed by this PR as well?

File: helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java [hunk 2]
```diff
@@ -85,7 +85,7 @@ public void onBecomeStandbyFromLeader(Message message, NotificationContext conte
 
     logger.info(controllerName + " becoming standby from leader for " + clusterName);
 
-    if (_controller != null) {
+    if (_controllerOpt.isPresent()) {
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java:88]: If this is not in the scope of lock, could we leave leakage there when a race condition happen?

File: helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java (hunk 1 above)
Comment (jiajunwang) [reply helix-core/src/test/java/org/apache/helix/monitoring/TestClusterStatusMonitorLifecycle.java:297]: That is not clear to me. I prefer to validate by keeping tracking unstable tests.

Moreover, the reason I remove issue 1280 is that it is an out-of-date comment. There is no clear relationship between that issue and the problem here.

File: helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java (hunk 2 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java:88]: 1. State transitions are all happening in one thread. So there is no race condition between the reset here and the creation logic. This PR is to fix the NPE when the reset is triggered by shutdown. So this line won't cause the issue even without the fix.
2. This fix is only to fix NPE, it won't fix all potential leakage. Note that if the partition is in ERROR state, then when we drop the partition, there is no clean up. This is the most serious leakage, actually. If the test is still not stable, I plan to fix this problem in another PR.

Comment (junkaixue) [reply helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java:88]: I remember Xiaoyuan's change will make the state model can call reset and ST at same time, right?

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/participant/DistClusterControllerStateModel.java:88]: That would be fine. Your concern is basically "onBecomeLeaderFromStandby" and "onBecomeStandbyFromLeader" are called at the same time. But even with Xiaoyuan's change, this is not possible, right?
If the reset is triggered by the other logic, like shutdown, then we have synchronized (_controllerOpt) in reset() to ensure the calls are mutually exclusive.

**Review by junkaixue — APPROVED** [review]

**Comment by jiajunwang** [comment]
Approved by @junkaixue , I will merge the code soon.

**Pull request merged**